component/file_cache/file_cache.go

```diff
@@ -453,22 +477,23 @@ func (fc *FileCache) StreamDir(options internal.StreamDirOptions) ([]*internal.O
 			// Enumerate over the results from the local cache and add to attrs
 			for _, entry := range dirents {
 				entryPath := common.JoinUnixFilepath(options.Name, entry.Name())
-				entryCachePath := common.JoinUnixFilepath(fc.tmpPath, entryPath)
-
-				info, err := os.Stat(entryCachePath) // Grab local cache attributes
-				// If local file is not locked then only use its attributes otherwise rely on container attributes
-				if err == nil && !info.IsDir() &&
-					!fc.fileLocks.Locked(entryPath) {
-
-					// This is an overhead for streamdir for now
-					// As list is paginated we have no way to know whether this particular item exists both in local cache
-					// and container or not. So we rely on getAttr to tell if entry was cached then it exists in cloud storage too
-					// If entry does not exists on storage then only return a local item here.
-					_, err := fc.NextComponent().GetAttr(internal.GetAttrOptions{Name: entryPath})
-					if err != nil && (err == syscall.ENOENT || os.IsNotExist(err)) {
-						log.Debug("FileCache::StreamDir : serving %s from local cache", entryPath)
-						attr := newObjAttr(entryPath, info)
-						attrs = append(attrs, attr)
+
+				if !entry.IsDir() && !fc.fileLocks.Locked(entryPath) {
+					entryCachePath := common.JoinUnixFilepath(fc.tmpPath, entryPath)
+					info, err := os.Stat(entryCachePath) // Grab local cache attributes
+					// If local file is not locked then only use its attributes otherwise rely on container attributes
+					if err == nil {
+						// This is an overhead for streamdir for now
+						// As list is paginated we have no way to know whether this particular item exists both in local cache
+						// and container or not. So we rely on getAttr to tell if entry was cached then it exists in cloud storage too
+						// If entry does not exists on storage then only return a local item here.
+						_, err := fc.NextComponent().GetAttr(internal.GetAttrOptions{Name: entryPath})
+						if err != nil && (err == syscall.ENOENT || os.IsNotExist(err)) {
+							// Case 2 (file only in local cache) so create a new attributes and add them to the storage attributes
+							log.Debug("FileCache::StreamDir : serving %s from local cache", entryPath)
+							attr := newObjAttr(entryPath, info)
```

Comment:
Since `info` is only used when `GetAttr` returns an error, can we fetch `info` here instead of doing so before calling `GetAttr`?

Reply:
Yes, good catch. This will be much faster.